src/split_at.rs

```diff
@@ -233,59 +213,834 @@ unsafe impl<T> SplitAt for [T] {
     }
 }
 
-/// Splits `T` in two.
+/// A `T` that has been split into two possibly-overlapping parts.
 ///
-/// # Safety
+/// For some dynamically sized types, the padding that appears after the
+/// trailing slice field [is a dynamic function of the trailing slice
+/// length](KnownLayout#slice-dst-layout). If `T` is split at a length that
+/// requires trailing padding, the trailing padding of the left part of the
+/// split `T` will overlap the right part. If `T` is a mutable reference or
+/// permits interior mutation, you must ensure that the left and right parts do
+/// not overlap. You can do this at zero-cost using using
+/// [`Self::via_immutable`], [`Self::via_into_bytes`], or
+/// [`Self::via_unaligned`], or with a dynamic check by using
+/// [`Self::via_runtime_check`].
 ///
-/// The caller promises that:
-/// 0. `l_len.get()` is not greater than the length of `ptr`'s trailing slice.
-/// 1. if `I::Aliasing` is [`Exclusive`] or `T` permits interior mutation, then
-///    `l_len.padding_needed_for() == 0`.
-#[inline(always)]
-unsafe fn ptr_split_at_unchecked<'a, T, I, R>(
-    ptr: Ptr<'a, T, I>,
-    l_len: MetadataOf<T>,
-) -> (Ptr<'a, T, I>, Ptr<'a, [T::Elem], I>)
+/// If is impossible to concurrently reference both parts, you can convert a
+/// `Split` into a reference to only the left part (via [`Self::left`] or to
+/// only right part (via [`Self::right`]).
+#[derive(Clone, Copy, Debug)]
```

Comment:
```suggestion
#[derive(Debug)]
```
The `Clone` and `Copy` impls are, in principle, unsound. Since a `Split` represents two possibly overlapping parts, it should not be possible to obtain concurrent borrows to both parts of the split by copying the `Split`.

Annoying, demonstrating the issue with a test case is seemingly impossible.  The only way to currently create a `Split<&T>` requires an `Immutable` bound, so the problem of overlapping mutable parts cannot be demonstrated with a `&T`. And the only way to currently create a `Split<&mut T>` requires an `IntoBytes` bound, which precludes the existence of trailing padding that would result in overlapping parts.

Reply:
Derives removed: https://github.com/google/zerocopy/compare/02b927ec93f33331e5e178bc8071db6d3bd366e0..135e87f6f3b6534727f99c3f0d09ab78a473eb1e#diff-394d6458153031257865a7aaddcda3ac55a4ae61c642dee241136b08a66fcf90L232